### PR TITLE
Disable RTTI on Unix

### PR DESCRIPTION
**Pull request by am11**

### CODE REVIEW

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -430,6 +430,8 @@ if (CLR_CMAKE_HOST_UNIX)
   # optimization in the IEEE 754 specification and is therefore considered unsafe.
   add_compile_options(-ffp-contract=off)
 
+  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fno-rtti>)
```
Comment (AaronRobinsonMSFT) [inline eng/native/configurecompiler.cmake:433]: I don't think we use this on Windows either. Can we try disabling there too?

Comment (am11) [reply eng/native/configurecompiler.cmake:433]: It is disabled on Windows: https://github.com/dotnet/runtime/blob/abfd53a204547b862b324ef5012d364916860c3c/eng/native/configurecompiler.cmake#L628-L631 IIRC, this replacement approach prevented a warning (related to `/GR` overriding `/GR-` or the other way around).

**Comment by am11** [comment]
@AaronRobinsonMSFT, (unrelated to this PR but) I just noticed gcc release build is broken in `main`:

```
$ cd runtime
$ docker run --rm -v$(pwd):/runtime mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-gcc12-amd64-20220511124845-b7a6185 \
    /runtime/src/coreclr/build-runtime.sh -gcc -c Release

...
[ 36%] Building CXX object utilcode/CMakeFiles/utilcode_obj.dir/fstream.cpp.o
/runtime/src/coreclr/vm/appdomain.cpp: In member function ‘void DomainLocalModule::EnsureDynamicClassIndex(DWORD)’:
/runtime/src/coreclr/vm/appdomain.cpp:4191:15: error: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of non-trivially copyable type ‘struct DomainLocalModule::DynamicClassInfo’; use copy-assignment or copy-initialization instead [-Werror=class-memaccess]
 4191 |         memcpy(pNewDynamicClassTable, m_pDynamicClassTable, sizeof(DynamicClassInfo) * oldDynamicEntries);
      |         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /runtime/src/coreclr/vm/common.h:312,
                 from /runtime/artifacts/obj/coreclr/linux.x64.Release/vm/wks/CMakeFiles/cee_wks_core.dir/cmake_pch.hxx:5,
                 from <command-line>:
/runtime/src/coreclr/vm/appdomain.hpp:164:12: note: ‘struct DomainLocalModule::DynamicClassInfo’ declared here
  164 |     struct DynamicClassInfo
      |            ^~~~~~~~~~~~~~~~
```

Most of these were fixed in https://github.com/dotnet/runtime/commit/53841055e735895520633ecf5da392ac14415ba8. The gcc leg in CI only covers checked builds and for some reason this only showed up in the Release build.

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT, (unrelated to this PR but) I just noticed gcc release build is broken in main:

Sigh... this is classic undefined behavior. I'm also unclear as to why this is only failing on a Release build. Both the `VolatilePtr` and `Volatile` types are particularly troublesome in this manner. Since we know this is "okay" feel free to add the casts to `(void*)` to the arguments above to quiet the error.

File: src/coreclr/vm/gdbjit.cpp [hunk 2]
```diff
@@ -1384,22 +1384,23 @@ bool FunctionMember::GetBlockInNativeCode(int blockILOffset, int blockILLen, TAD
...
 void FunctionMember::DumpTryCatchBlock(char* ptr, int& offset, int ilOffset, int ilLen, int abbrev)
 {
-    TADDR startOffset;
-    TADDR endOffset;
+    TADDR* startOffset;
+    TADDR* endOffset;
 
-    if (!GetBlockInNativeCode(ilOffset, ilLen, &startOffset, &endOffset))
+    if (!GetBlockInNativeCode(ilOffset, ilLen, startOffset, endOffset))
         return;
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/gdbjit.cpp:1391]: This doesn't look correct. These pointers aren't initialized and are being passed to a function that is dereferencing them. They are also being dereferenced below.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by am11** [comment]
Binary size reduction:

```diff
--- main
+++ PR
@@ -1,29 +1,29 @@
-812648 /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrgc.so
+811592 /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrgc.so
 
-2740384        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit.so
+2727936        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit.so
 
-2690048        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_universal_arm64_x64.so
+2677040        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_universal_arm64_x64.so
 
-2353648        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_universal_arm_x64.so
+2340688        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_universal_arm_x64.so
 
-2740408        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_unix_x64_x64.so
+2727960        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_unix_x64_x64.so
 
-2720312        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_win_x64_x64.so
+2707864        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_win_x64_x64.so
 
-2679872        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_win_x86_x64.so
+2668144        /runtime/artifacts/bin/coreclr/linux.x64.Release/libclrjit_win_x86_x64.so
 
-6971728        /runtime/artifacts/bin/coreclr/linux.x64.Release/libcoreclr.so
+6912272        /runtime/artifacts/bin/coreclr/linux.x64.Release/libcoreclr.so
 
 877784 /runtime/artifacts/bin/coreclr/linux.x64.Release/libcoreclrtraceptprovider.so
 
-48736  /runtime/artifacts/bin/coreclr/linux.x64.Release/libjitinterface_x64.so
+48128  /runtime/artifacts/bin/coreclr/linux.x64.Release/libjitinterface_x64.so
 
-2446384        /runtime/artifacts/bin/coreclr/linux.x64.Release/libmscordaccore.so
+2412992        /runtime/artifacts/bin/coreclr/linux.x64.Release/libmscordaccore.so
 
-1686416        /runtime/artifacts/bin/coreclr/linux.x64.Release/libmscordbi.so
+1627376        /runtime/artifacts/bin/coreclr/linux.x64.Release/libmscordbi.so
 
-1264600        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-collector.so
+1256424        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-collector.so
 
-1271088        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-counter.so
+1262928        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-counter.so
 
-1256800        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-simple.so
+1248656        /runtime/artifacts/bin/coreclr/linux.x64.Release/libsuperpmi-shim-simple.so

**Pull request merged**